Software

T2heke opened the pull request:
Edited bootcamp code.

ReikoR committed:
Initialize software

ReikoR committed:
Initialize software

T2heke committed:
modified:   picr22-boot-camp-programming (modified content, untracked content)

T2heke committed:
modified:   picr22-boot-camp-programming (modified content, untracked content)

T2heke committed:
modified:   picr22-boot-camp-programming

T2heke committed:
modified:   picr22-boot-camp-programming

T2heke committed:
new file:   Color.py
	new file:   README.md
	new file:   __pycache__/Color.cpython-39.pyc
	new file:   __pycache__/camera.cpython-310.pyc
	new file:   __pycache__/camera.cpython-39.pyc
	new file:   __pycache__/image_processor.cpython-310.pyc
	new file:   __pycache__/image_processor.cpython-39.pyc
	new file:   __pycache__/motion.cpython-39.pyc
	new file:   camera.py
	new file:   colors/colors.pkl
	new file:   config_colors.py
	new file:   image_processor.py
	new file:   main.py
	new file:   motion.py
	new file:   segment_module/Segment.egg-info/PKG-INFO
	new file:   segment_module/Segment.egg-info/SOURCES.txt
	new file:   segment_module/Segment.egg-info/dependency_links.txt
	new file:   segment_module/Segment.egg-info/top_level.txt
	new file:   segment_module/build/lib.linux-x86_64-cpython-39/segment.cpython-39-x86_64-linux-gnu.so
	new file:   segment_module/build/temp.linux-x86_64-cpython-39/segment.o
	new file:   segment_module/colors/colors.pkl
	new file:   segment_module/segment.c
	new file:   segment_module/setup.py
	new file:   segment_module/test.py

T2heke committed:
new file:   Color.py
	new file:   README.md
	new file:   __pycache__/Color.cpython-39.pyc
	new file:   __pycache__/camera.cpython-310.pyc
	new file:   __pycache__/camera.cpython-39.pyc
	new file:   __pycache__/image_processor.cpython-310.pyc
	new file:   __pycache__/image_processor.cpython-39.pyc
	new file:   __pycache__/motion.cpython-39.pyc
	new file:   camera.py
	new file:   colors/colors.pkl
	new file:   config_colors.py
	new file:   image_processor.py
	new file:   main.py
	new file:   motion.py
	new file:   segment_module/Segment.egg-info/PKG-INFO
	new file:   segment_module/Segment.egg-info/SOURCES.txt
	new file:   segment_module/Segment.egg-info/dependency_links.txt
	new file:   segment_module/Segment.egg-info/top_level.txt
	new file:   segment_module/build/lib.linux-x86_64-cpython-39/segment.cpython-39-x86_64-linux-gnu.so
	new file:   segment_module/build/temp.linux-x86_64-cpython-39/segment.o
	new file:   segment_module/colors/colors.pkl
	new file:   segment_module/segment.c
	new file:   segment_module/setup.py
	new file:   segment_module/test.py

T2heke committed:
deleted:    picr22-boot-camp-programming

T2heke committed:
deleted:    picr22-boot-camp-programming

T2heke committed:
new file:   software/bootcamp-code/.idea/.gitignore
	new file:   software/bootcamp-code/.idea/bootcamp-code.iml
	new file:   software/bootcamp-code/.idea/inspectionProfiles/profiles_settings.xml
	new file:   software/bootcamp-code/.idea/misc.xml
	new file:   software/bootcamp-code/.idea/modules.xml
	new file:   software/bootcamp-code/.idea/vcs.xml

T2heke committed:
new file:   software/bootcamp-code/.idea/.gitignore
	new file:   software/bootcamp-code/.idea/bootcamp-code.iml
	new file:   software/bootcamp-code/.idea/inspectionProfiles/profiles_settings.xml
	new file:   software/bootcamp-code/.idea/misc.xml
	new file:   software/bootcamp-code/.idea/modules.xml
	new file:   software/bootcamp-code/.idea/vcs.xml

T2heke committed:
modified:   software/README.asciidoc
	modified:   software/bootcamp-code/main.py

T2heke committed:
modified:   software/README.asciidoc
	modified:   software/bootcamp-code/main.py

ReikoR committed:
Merge branch 'software' of github.com:ut-robotics/picr22-team-twelve into software

T2heke committed:
modified:   software/bootcamp-code/main.py

T2heke committed:
Merge branch 'software' of github.com:ut-robotics/picr22-team-twelve into software
yes

T2heke committed:
Update main.py

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -0,0 +1,122 @@
+            print("Serial port assigned")
+            break
+
+    omni_motion.open(port)
```
Determining the port should happen in the motion class and should not accept a port as an argument.

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -0,0 +1,122 @@
+            #find time passed since the start of program
+            time_passed = time.time()-start_time
+            
+            move_wheel_s(omni_motion, 0, 2, time_passed, 20)
```
I am not quite sure what this logic does. Overall I would suggest creating a state machine where a variable holds the current behavior state of the robot as an enum (Refer to Color.py from the boot camp code), for example "FIND_BALL" or "DRIVE_TO_BALL" and based on the value, it selects the correct logic block from a if-elif-else block inside the main loop.

This allows to iterate on specific behavior logic parts and makes it simpler to avoid code changes affecting the logic in unknown ways.

Akustav reviewed (changes requested):
The main change required is extending behavior logic in the main loop. I would suggest implementing a state machine there.

T2heke committed:
Update motion.py

T2heke committed:
Update main.py

T2heke committed:
Update main.py

T2heke committed:
Update motion.py

T2heke committed:
Update main.py

T2heke committed:
Update motion.py

Started on the moving logic.

T2heke committed:
new file:   .gitignore
	modified:   bootcamp-code/__pycache__/Color.cpython-39.pyc
	modified:   bootcamp-code/__pycache__/camera.cpython-310.pyc
	modified:   bootcamp-code/__pycache__/camera.cpython-39.pyc
	modified:   bootcamp-code/__pycache__/image_processor.cpython-310.pyc
	modified:   bootcamp-code/__pycache__/image_processor.cpython-39.pyc
	modified:   bootcamp-code/__pycache__/motion.cpython-39.pyc

T2heke committed:
modified:   __pycache__/motion.cpython-39.pyc
	modified:   main.py
	modified:   motion.py

T2heke committed:
modified:   .gitignore

T2heke committed:
deleted:    .idea/.gitignore

T2heke committed:
modified:   main.py

T2heke committed:
modified:   colors/colors.pkl
	modified:   main.py
	modified:   motion.py

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -0,0 +1,137 @@
+            # objects of interest and calculate the required motion for reaching the objects
+            print()
+
+            if state>1:
```
I suggest making states control state transitions, then all logic is consolidated in one place. The robot can always start with `find ball` and if it instantly finds a ball, the find ball state sets the new state to `drive to ball`. If that state has no balls to drive to, it will transition back to finding a ball.

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -0,0 +1,137 @@
+
+            # STATE TO FIND THE BALL
+            if state == 0:
+                largest_ball_xco = 0
+                largest_ball_yco = 0
```
Not sure why this is needed. Maybe just give the move method 0-s as x and y speed inputs

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -0,0 +1,137 @@
+                # the biggest ball is the first one in the detected balls list
+                # the coordinates of the largest ball, THE ZERO COORDINATES OF THE FRAME ARE IN THE UPPER LEFT CORNER
+                # Using these coordinates calculate the side speed, forward speed and rotation for the robot
+                largest_ball_Object = processedData.balls[0]
```
variable naming should follow a uniform styling. If one can not decide what to use, I suggest using PEP 8 as a guide: https://peps.python.org/pep-0008/

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -0,0 +1,137 @@
+                # Using these coordinates calculate the side speed, forward speed and rotation for the robot
+                largest_ball_Object = processedData.balls[0]
+                # We want the ball to end up in the middle of the frame
+                ball_desired_x = cam.rgb_width/2
```
Constants like this should be calculated before, not in every loop.

Akustav commented on software/bootcamp-code/motion.py:
```diff
@@ -0,0 +1,156 @@
+        robotDirectionAngle = math.atan2(y_speed, x_speed)
+        # robot angular speed is basically robot rotational speed
+        robotAngularSpeed = rot_speed
+        for s in speeds:
```
for s in speeds returns the elements of the list, no?
![image](https://user-images.githubusercontent.com/5435954/195283084-4112489b-036c-4a31-b7ac-0815aa2b2d02.png)
You most seem to use want to use the list indexes. In this case you can use the `for i in range(3)` 
![image](https://user-images.githubusercontent.com/5435954/195283446-93b9c492-b170-4b3b-8cf8-27da6b0cc5d5.png)

Akustav reviewed (commented):
State machine needs to be migrated to enum and little tweaks with state changing.

Also check the motor speed calculation

T2heke committed:
Update main.py

T2heke committed:
Update motion.py

T2heke committed:
Update main.py

T2heke committed:
Update motion.py

T2heke committed:
Update main.py

Akustav reviewed (approved):
There are some variable naming inconsistencies but the code seems OK
